// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.simulator";
option java_outer_classname = "SimulatorProto";
option objc_class_prefix = "SIM";

import "google/protobuf/empty.proto";
//import "ndarray.proto";
package simulator;

// Interface exported by the server.
service SimulatorServer {

  rpc GetSimulationConfigMessage(google.protobuf.Empty) returns (SimulationConfig) {}
  rpc GetSimulationConfigSerialized(google.protobuf.Empty) returns (SimulationConfigSerialized) {}
  rpc GetRecordedChanges(Name) returns (SerializedDict) {}

  rpc GetAgentConfigMessage(google.protobuf.Empty) returns (AgentConfig) {}
  rpc GetAgentConfigSerialized(google.protobuf.Empty) returns (AgentConfigSerialized) {}

  rpc GetPopulationConfigMessage(google.protobuf.Empty) returns (PopulationConfig) {}
  rpc GetPopulationConfigSerialized(google.protobuf.Empty) returns (PopulationConfigSerialized) {}

  rpc SetSimulationConfigSerialized(SimulationConfigSerialized) returns (google.protobuf.Empty) {}
  rpc SetSimulationConfig(SimulationConfig) returns (google.protobuf.Empty) {}


  rpc SetPopulationConfig(PopulationConfig) returns (google.protobuf.Empty) {}
  rpc SetPopulationConfigSerialized(PopulationConfigSerialized) returns (google.protobuf.Empty) {}

  rpc GetStateMessage(google.protobuf.Empty) returns (State) {}

  rpc GetStateArrays(google.protobuf.Empty) returns (StateArrays) {}

  rpc IsStarted(google.protobuf.Empty) returns (IsStartedState) {}

  rpc Start(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc Stop(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc UpdateNeighborFn(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc UpdateStateNeighbors(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc UpdateFunctionUpdate(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc UpdateBehaviors(google.protobuf.Empty) returns (google.protobuf.Empty) {}

  rpc GetChangeTime(google.protobuf.Empty) returns (Time) {}

}

message Name {
  string name = 1;
}

message Time {
  int32 time = 1;
}

message SimulationConfig {
  float box_size = 1;
  int32 map_dim = 2;

  int32 num_steps_lax = 3;
  int32 num_lax_loops = 4;
  float dt = 5;
  float freq = 6;
  bool to_jit = 7;
  int32 n_agents = 8;
  NDArray entity_behaviors = 9;
}

message SimulationConfigSerialized {
  string serialized = 1;
}

message SerializedDict {
  string serialized_dict = 1;
  bool has_entity_behaviors = 2;
  NDArray entity_behaviors = 3;
}

message AgentConfig {
    float wheel_diameter = 1;
    float base_length = 2;
    float speed_mul = 3;
    float theta_mul = 4;
    float neighbor_radius = 5;
    float proxs_dist_max = 6;
    float proxs_cos_min = 7;
}

message AgentConfigSerialized {
  string serialized = 1;
}

message PopulationConfig {
  int32 n_agents = 1;
}

message PopulationConfigSerialized {
  string serialized = 1;
}

message Position {
  repeated float x = 1;
  repeated float y = 2;
}

message State {
  Position positions = 1;
  repeated float thetas = 2;
}

message NDArray {
  bytes ndarray = 1;
}

message StateArrays {
  NDArray positions = 1;
  NDArray thetas = 2;
  NDArray proxs = 3;
  NDArray motors = 4;
  int32 entity_type = 5;
}

message IsStartedState {
  bool is_started = 1;
}
